### DISCUSSION TITLE
Tree shaking of packages

### DISCUSSION THREAD

**Discussion by vimtor** [vercel/turborepo#1637] [Help] [answered]
I've noticed that by default the whole package is bundled into the application that imports it. If for example I have a `docs` Next.js app and `ui` package, and I import `Button` in one page, the whole bundle is included.

I understand that this is not the scope of Turborepo. In multi-repo approach you will still have to use bundler like [`tsup`](https://www.npmjs.com/package/tsup) to include the necessary tree shaking annotations, mark the `package.json` with `"sideEffects": false`, etc.

I've had some problems with this. I've tried to follow the examples provided for the UI where some bundlers are used but it seems to not be working. I keep wondering if I am missing something. Maybe more documentation could be provided or you could point to some resource to learn about the topic.

Besides that, I love Turborepo, it has made my development experience much better ❤️

**Comment by alexn-s** [comment]
it would be neat for turbo to handle that, but i dont think it is feasible for turbo to provide the feature or more information, because of the scope of the task.  there are too many paths a turbo user could take (pnpm workspaces, package projects, yarn, ...). i think it would only make sense for a solution like lerna/ nx that are heavily opinionated

what monorepo solution do you use (which toolings)?

**Reply by vimtor** [reply]
I am using Turborepo with NPM workspaces and I am really happy about it. I followed the example provided [here](https://github.com/vercel/turborepo/tree/main/examples/design-system) but the `utils` package I created keeps importing the whole package instead of individual functions and I don't know where to start with for solving the issue.

**Reply by EarlGeorge** [reply]
https://tsup.egoist.dev **_use it._**

**Comment by anthonyshew** [comment, marked as answer]
We've updated [the examples directory](https://github.com/vercel/turbo/tree/main/examples) to stop using barrel files, supporting proper tree shaking with Internal Packages.
